src/sinks/util/test.rs

```diff
@@ -106,3 +111,37 @@ where
 
     (rx, trigger, server)
 }
+
+pub async fn get_received_gzip(
+    rx: mpsc::Receiver<(Parts, Bytes)>,
+    assert_parts: impl Fn(Parts),
+) -> Vec<String> {
+    rx.flat_map(|(parts, body)| {
+        assert_parts(parts);
+        stream::iter(BufReader::new(MultiGzDecoder::new(body.reader())).lines())
+    })
+    .map(Result::unwrap)
+    .map(|line| {
+        let val: serde_json::Value = serde_json::from_str(&line).unwrap();
+        val.get("message").unwrap().as_str().unwrap().to_owned()
+    })
+    .collect::<Vec<_>>()
+    .await
+}
+
+pub async fn get_received_zlib(
+    rx: mpsc::Receiver<(Parts, Bytes)>,
+    assert_parts: impl Fn(Parts),
+) -> Vec<String> {
+    rx.flat_map(|(parts, body)| {
+        assert_parts(parts);
+        stream::iter(BufReader::new(ZlibDecoder::new(body.reader())).lines())
+    })
+    .map(Result::unwrap)
+    .map(|line| {
+        let val: serde_json::Value = serde_json::from_str(&line).unwrap();
+        val.get("message").unwrap().as_str().unwrap().to_owned()
+    })
+    .collect::<Vec<_>>()
+    .await
+}
```

Comment:
<!-- Thoughts represent an idea that popped up from reviewing. These comments are non-blocking by nature, but they are extremely valuable and can lead to more focused initiatives and mentoring opportunities.  -->
**💭 thought:** ‏On the surface at least, it looks like these could be collapsed to accept a decoder as a parameter. Is there a reason not to do this?

Reply:
I had chosen to do this as opposed to instantiating a new decoder with each test. What do you think?

Reply:
One approach could be to have the helper that takes the decoder as a parameter, and then still have `get_received_[gzip/zlib]()` , those two just create the decoder and call the helper.

Reply:
Looked into this and faced some challenges. The key one being that decoders like GzDecoder and ZlibDecoder do not share a common trait out of the box that we can use for typing a function parameter expecting any "decoder". A solution is not obvious to me, any advice?

Reply:
Sike, I found a solution using closures 🥳